### PR TITLE
32 build binary classifier pipeline including wandb

### LINKED ISSUE #32: Build binary classifier pipeline including wandb

### PR DESCRIPTION
**Pull request by RFOxbury**
---

# Description

- Creates dummy data using langchain
- Evaluates a dummy classifier (logged on wandb)
- Creates jobbert embeddings for the dummy data and trains a logistic regression model (logged on wandb)

Fixes #32 

# Instructions for Reviewer

In order to test the code in this PR you need to run:

- [ ] `dap_job_quality/pipeline/sentence_classifier/prep_training_data.py`
- [ ] `dap_job_quality/pipeline/sentence_classifier/baseline_classifier.py`
- [ ] `dap_job_quality/pipeline/sentence_classifier/log_reg.py`

Please pay special attention to ...

# Checklist:

- [ ] I have refactored my code out from `notebooks/`
- [ ] I have checked the code runs
- [ ] I have tested the code
- [ ] I have run `pre-commit` and addressed any issues not automatically fixed
- [ ] I have merged any new changes from `dev`
- [ ] I have documented the code
  - [ ] Major functions have docstrings
  - [ ] Appropriate information has been added to `README`s
- [ ] I have explained this PR above
- [ ] I have requested a code review

### CODE REVIEW

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+    for data, name in zip(
+        [X_train, X_val, X_test, y_train, y_val, y_test],
+        ["X_train", "X_val", "X_test", "y_train", "y_val", "y_test"],
+    ):
+        save_to_s3(
+            BUCKET_NAME,
+            data,
+            f"job_quality/sentence_classifier/inputs/labelled/{name}_dummy.pkl",
+        )
```
Comment (cmbrennan002) [inline dap_job_quality/pipeline/sentence_classifier/prep_training_data.py:96]: This is awesome! I did get a warning after running this file - do you get the same?  Since it's a data prep step, I think it's fine - but wanted to check. 

`Some weights of BertModel were not initialized from the model checkpoint at jjzha/jobbert-base-cased and are newly initialized: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.`

File: dap_job_quality/pipeline/sentence_classifier/baseline_classifier.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    cm = confusion_matrix(y_val, y_pred)
+    cm = pd.DataFrame(cm)
+    logging.info(cm)
+
+    # Log confusion matrix
+    wb_confusion_matrix = wandb.Table(data=cm, columns=["0", "1"])
+    run.log({"confusion_matrix": wb_confusion_matrix})
+
+    # End the weights and biases run
+    wandb.finish()
```
Comment (cmbrennan002) [inline dap_job_quality/pipeline/sentence_classifier/baseline_classifier.py:53]: This looks great, and ran really well! And thanks for setting me up on W&B - very cool.

File: dap_job_quality/pipeline/sentence_classifier/generate_dummy_data.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+OUT_FILE = PROJECT_DIR / f"inputs/labelled/{FILENAME}"
+OUT_FILE.parent.mkdir(parents=True, exist_ok=True)
+
+load_dotenv()
+OPENAI_API_KEY = os.environ.get("OPENAI_API_KEY")
+
+INPUT = """
+
+Generate a job advert sentence and return it with a label if it is related to job quality or not. The label will be 1 if the sentence relates to the compensation package, the contract type, working hours, info about the team you'll be joining or sentences that relate to “why you should want to work here”; it will be 0 if the sentence describes role requirements, "what will be expected of you", "how to apply", or anything else that does not directly determine job quality.
+
```
Comment (cmbrennan002) [inline dap_job_quality/pipeline/sentence_classifier/generate_dummy_data.py:32]: I really like this prompt. This might just be a matter of labelling data, but from the manual labelling, a huge majority of skill/ benefit confusion is around training/ l&d (whether the job offers training, or requires you to run/ offer training to others). Should we add something in to ensure enough representation of that distinction in the training data? Or you think it will naturally occur with a big enough sample?

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+from dap_job_quality import PROJECT_DIR, config, BUCKET_NAME
+from dap_job_quality.getters.labelled_data import get_dummy_job_sentences
+from dap_job_quality.getters.data_getters import save_to_s3
+
+SENT_MODEL = config["sentence_model"]
+SEED = 42
+
+
+def embed_sentences(sentences, model_name=SENT_MODEL):
```
Comment (cmbrennan002) [inline dap_job_quality/pipeline/sentence_classifier/prep_training_data.py]: Could you add a doc string here?

File: dap_job_quality/pipeline/sentence_classifier/baseline_classifier.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    run = wandb.init(
+        project="dap-job-quality",
+        job_type="Sentence classifier",
+        save_code=True,
+        tags=[f"baseline_model"],
+    )
+
+    label_probabilities = y_train.value_counts(normalize=True)
+
+    y_pred = np.random.choice(
```
Comment (cmbrennan002) [inline dap_job_quality/pipeline/sentence_classifier/baseline_classifier.py:35]: I really like this, it's very thorough. Because it's a small sample, however, I think the output is pretty volatile (eg, I ran it twice, and got an f_1 score of .78 once, and .69 once). If we stay with a fairly small sample of labelled data, should we change the code to run this say 100 (or maybe 1000 times) and take the mean? So it's a more stable baseline?

File: dap_job_quality/pipeline/sentence_classifier/log_reg.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+    run = wandb.init(
+        project="dap-job-quality",
+        job_type="Sentence classifier",
+        save_code=True,
+        tags=[f"logistic_regression"],
+    )
+
+    # Train a logistic regression model
+    model = LogisticRegression(max_iter=1000)
```
Comment (cmbrennan002) [inline dap_job_quality/pipeline/sentence_classifier/log_reg.py]: This looks great, and I think good to go as an initial pipeline. As a next step (feel free to disregard for this PR!) my only concern here would be over-fitting (particularly with such a small validation/ test set):
- Should we test how this goes with fewer iterations? (10 x more iterations than sample size seems like a lot to me?) 
- Should we try adding a penalty term? 
- Should we set a random seed?

**Comment by cmbrennan002** [comment]
Hey Rosie,

I think this looks really great! I've left some comments on the files. And cool to see pretty high accuracy off a relatively simple method. Will be interesting to see how it goes with the labelled data.

Also - very cool to see wanb! I had never used it before, seems like a really useful tool. 

Clare

File: dap_job_quality/pipeline/sentence_classifier/prep_training_data.py (hunk 1 above)
Comment (RFOxbury) [reply dap_job_quality/pipeline/sentence_classifier/prep_training_data.py:96]: This is a good point and tbh I'm not too sure, so I've messaged Liz to see what she thinks!

File: dap_job_quality/pipeline/sentence_classifier/baseline_classifier.py (hunk 5 above)
Comment (RFOxbury) [reply dap_job_quality/pipeline/sentence_classifier/baseline_classifier.py:35]: This is a really good point. For simplicity I've just added a random seed at the top of the script so that the volatility is gone.
